cpp/include/raft/neighbors/detail/cagra/compute_distance.hpp

```diff
@@ -252,17 +255,22 @@ struct standard_dataset_descriptor_t
             // because:
             // - Above the last element (dataset_dim-1), the query array is filled with zeros.
             // - The data buffer has to be also padded with zeros.
-            DISTANCE_T diff = query_ptr[device::swizzling(kv)];
-            diff -= spatial::knn::detail::utils::mapping<float>{}(dl_buff[e].val.data[v]);
-            norm2 += diff * diff;
+            DISTANCE_T d = query_ptr[device::swizzling(kv)];
+            if (metric == raft::distance::L2Expanded) {
```

Comment:
It would be better to move this conditional branch outside of the loop so that it is not evaluated at every iteration and we can get better throughput.

Reply:
Made the metric a tempate arg. Will make it easier to add other dist metrics in the future.

Reply:
The template arg is better from a performance perspective, but the current runtime dispatch also seems acceptable. I have checked that the throughput of the L2-norm search is almost the same as the branch-24.06 version.

Reply:
Many thanks @enp1s0 for confirming this!